File: JavaScript/DataStructures/polyTreeNode.js
class PolyTreeNode {
    constructor(value) {
        this.value = value;
        this.parent = null;
        this.children = [];
    }

    parental() {
        return this.parent
    }

    valueOf() {
        return this.value
    }

    childrenOf() {
        return this.children.slice();
    }

    privateChildren() {
        return this.children
    }

    set childrenArray(arr) {
        this.children = arr
    }

    set parentEquals(parent) {
        if(this.parent === parent) { return };

        if(this.parent) {
            let kids = this.parent.privateChildren();
            let arr = [];
            for(let i = 0; i < kids.length; i++) {
                if(kids[i] === this) {
                    kids[i] = "";
                    continue;
                } else {
                    arr.push(kids[i])
                }
            }
            this.parent.childrenArray = arr;
        }
        this.parent = parent;
        parent ? this.parent.children.push(this) : null;
    }

    addChild(child) {
        child.parentEquals = this;
    }

    removeChild(child) {
        if(child) {
            child.parentEquals(nil);
        } else {
            throw "Not a child."
        }
    }

    dfs(target_value) {
        if((this.value[0] === target_value[0]) && (this.value[1] === target_value[1])) { return this };
        let kids = this.childrenOf();
        let result;
        for(let i = 0; i < kids.length; i++) {
            result = kids[i].dfs(target_value);
            if(result){ return result }
        }
        return null
    }

    bfs(target_value) {
        let opps = [this];
        while(opps.length !== 0) {
            let first = opps.shift();
            let kids = first.childrenOf();
            if((first.value[0] === target_value[0]) && (first.value[1] === target_value[1])){ return first }
            for(let i = 0; i < kids.length; i++) {
                opps.push(kids[i]);
            }
        }
        return null
    }
}

let btd = new PolyTreeNode([2,2]);
let walk = new PolyTreeNode([7,7]);
let cry = new PolyTreeNode([2,4]);

// cry.parentEquals = btd;
// console.log(walk)
// walk.parentEquals = btd;
// console.log(walk)
// btd.addChild(walk);
// btd.addChild(cry);
// console.log(btd.bfs("insane"));
// console.log(btd.childrenOf());
// console.log(btd.bfs([7,7]));

module.exports = PolyTreeNode;

File: RUBY/DataStructures/polyTreeNode.rb
class PolyTreeNode
    attr_reader :parent
    attr_accessor :value

    def initialize(value)
        @value = value
        @parent = nil
        @children = []
    end

    def parent
        @parent
    end

    def children
        @children.dup
    end

    def value
        @value
    end

    def parent=(parent)
        return if self.parent == parent

        if self.parent
            self.parent._children.each_with_index do |child, i|
                if child == self
                    self.parent._children[i] = ""
                    arr = self.parent._children.select{|ele| ele != ""}
                    self.parent._children = arr
                end
            end
        end

        @parent = parent
        self.parent._children << self unless self.parent.nil?
    end

    def add_child(child_node)
        child_node.parent=(self)
    end

    def remove_child(child_node)
        if child_node && self._children.include?(child_node)
            child_node.parent=(nil)
        else
            raise "Node is not a child"
        end
    end

    def dfs(target_value)
        return self if self.value == target_value
        self.children.each do |node|
            result = node.dfs(target_value)
            return result unless result.nil?
        end
        nil
    end

    def bfs(target_value)
        exposing = [self]
        while exposing.length != 0
            first = exposing.shift
            return first if first.value == target_value
            first.children.each do |child|
                exposing << child
            end
        end
        nil
    end
    
    def _children
        @children
    end

    protected

    def _children=(children)
        @children = children
    end
end

# btd = PolyTreeNode.new("root")
# walk = PolyTreeNode.new("insane")
# cry = PolyTreeNode.new("laugh")

# walk.parent = btd
# # p btd
# cry.parent = btd
# btd.remove_child(walk)
# p btd._children

